patches/server/0731-Add-System.out.println-catcher.patch

```diff
@@ -0,0 +1,106 @@
++import java.io.PrintStream;
++
++public final class SysoutCatcher {
++    private static final String OUT_PREFIX = "[STDOUT] ";
++    private static final String ERR_PREFIX = "[STDERR] ";
++    private final StackWalker stackWalker = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
++
++    public SysoutCatcher() {
++        // Listen to System.out.println
++        PrintStream wrappedOut = new PrintStream(System.out) {
++            @Override
++            public void println(String line) {
++                Class<?> clazz = stackWalker.getCallerClass();
++                try {
++                    JavaPlugin plugin = JavaPlugin.getProvidingPlugin(clazz);
++
++                    // Instead of just printing the message, send it to the plugin's logger
++                    plugin.getLogger().info(OUT_PREFIX + line);
++                    nag(plugin);
++                } catch (IllegalArgumentException e) {
++                    // If anything happens, the calling class doesn't exist, there is no JavaPlugin that "owns" the calling class, etc
++                    // Just print out normally, with some added information
++                    Bukkit.getLogger().info(String.format("[%s] %s %s", OUT_PREFIX, clazz.getName(), line));
++                }
++            }
++        };
++
++        // Listen to System.err.println
++        PrintStream wrappedErr = new PrintStream(System.err) {
++            @Override
++            public void println(String line) {
++                Class<?> clazz = stackWalker.getCallerClass();
++                try {
++                    JavaPlugin plugin = JavaPlugin.getProvidingPlugin(clazz);
++
++                    // Instead of just printing the message, send it to the plugin's logger
++                    plugin.getLogger().severe(ERR_PREFIX + line);
++                    nag(plugin);
++                } catch (IllegalArgumentException e) {
++                    // If anything happens, the calling class doesn't exist, there is no JavaPlugin that "owns" the calling class, etc
++                    // Just print out normally, with some added information
++                    Bukkit.getLogger().info(String.format("[%s] %s %s", ERR_PREFIX, clazz.getName(), line));
++                }
++            }
++        };
++
++        // Set the two PrintStreams so they're used
++        System.setOut(wrappedOut);
++        System.setErr(wrappedErr);
++    }
++
++    private void nag(Plugin plugin) {
++        if (isBlacklisted(plugin.getName())) return;
++        Bukkit.getLogger().warning(String.format("Nag author(s): '%s' of '%s' about their usage of System.out/err.print. " +
++                "Please use your plugin's logger instead (JavaPlugin#getLogger).",
++            plugin.getDescription().getAuthors(),
++            plugin.getName()));
++    }
++
++    private boolean isBlacklisted(final String pluginName) {
++        final String nags = System.getProperty("io.papermc.paper.suppress.sout.nags");
```

Comment:
It’s not gonna change; parse to a set and use that

Reply:
Done.